### DISCUSSION TITLE
Inline model documentation + tests

### DISCUSSION THREAD

**Discussion by jakebiesinger** [dbt-labs/dbt-core#6853] [Ideas]
The high-level requirement in my mind is "let me specify all the informational bits about a model inside the model itself". For me, this would be:
- model descriptions (`docs` blocks)
- column descriptions
- model + column tests

This topic seems rather perennial-- it keeps popping up again and again! Here's a quick citation list:
- This discussion https://github.com/dbt-labs/dbt-core/discussions/5093 which covers several approaches
- I wrote a series of PRs 4 years ago to allow `{% docs %}` blocks to be specified in `.sql` files [#1042](https://github.com/dbt-labs/dbt-core/pull/1042) and then to auto-attach them to models + columns based on the doc block name [#1043](https://github.com/dbt-labs/dbt-core/pull/1043)
- FR for the same in https://github.com/dbt-labs/dbt-core/issues/979, with lots of community traction + interactions
- Several scripts + external packages that work around this shortcoming
  - @vergenzt 's internal parsing script cited above
  - @HarlanH's .sql parser --> .md generator https://gist.github.com/HarlanH/277006989774372515c7130e63809315
  - https://github.com/dbt-labs/dbt-core/issues/1051 requesting that tests be specifiable inline (side point on the main FR there)
  - DBT discourse post on doing exactly this https://discourse.getdbt.com/t/here-is-a-way-to-write-dbt-docs-as-sql-comment/1658 and the associated python package https://github.com/anelendata/dbt_docstring (again, auto-generates yml files from sql files)
  - dbt-codegen https://github.com/dbt-labs/dbt-codegen, a macro intended to be placed in model .sql files and which dumps yml to stdout so you can manually copy-paste it into schema.yml files

It seems clear that the community wants this feature in some form. Why not bring it into `dbt-core` and give the people what they want? 😸 

Over the years, the DBT team has seemed to like this idea, but has raised concerns:
- The model file might get too big (dozens of columns in a model might be confusing, especially if all the docs are inlined)
- Concerns around availability of macros + properties within these blocks, due to the separate lifecycle DBT is running under the hood for documentation
- Desire for it to be clean + sufficiently explicit to not feel too magical / confusing
- Compatibility with existing systems, and having Yet Another Way to specify docs + schema

So far, we've seen proposals for syntax including:
- expanding the `config` jinja call to allow docs + descriptions + tests to be specified inline (so a single massive call to `config` would put all the things in place)
- Allow `docs` blocks in sql files, with possible support to auto-attach them to models (since they appear in the model sql file anyway)
- Custom "javadoc"-style annotations, possibly surrounded by `/** */` blocks (which can help with IDE confusion around jinja blocks) 

I am happy to (again) put in some elbow grease to get this done. Is there appetite for receiving it on the DBT team side?

**Comment by jtcohen6** [comment]
@jakebiesinger This was very cool to read. Thank you for kicking it off as a discussion, for having done your homework, and for sticking around through the years to see this idea through :)

There are two challenges here:
1. Finding a syntax that we can live with, given that we're ultimately talking about inlining structured data (currently expected as yaml) inside another language (Jinja-SQL, now also Python!)
2. Finding the capacity/priority to get this shipped. No question, this would be a nontrivial improvement to UX & ergonomics—as you note, it's been requested & upvoted by dozens of community members—and it would also require a nontrivial lift. Over the past few years, it's been tough for us to prioritize ergonomic improvements, relative to long-overdue needs around performance and foundational APIs.

It sounds like you'd be open to diving in and working on this, if we can set you up with a clear set of foundations to work in. Our parsing logic is still quite thorny (though better than it used to be); I couldn't promise a walk in the park.

I'm getting ahead of myself, though. First thing we'd need is an agreeable answer to (1).

## Finding a syntax

Let's check each of the options:

> expanding the `config` jinja call to allow docs + descriptions + tests to be specified inline (so a single massive call to config would put all the things in place)

I'm skeptical of this for two reasons:
- `config` has special rules around hierarchical inheritance / merging, which makes sense for most of its current options, and makes a lot less sense for the "non-config" properties (`description`, `tests`, `columns`).
- I don't think defining structured data in Jinja and passing it into the `{{ config() }}` macro is a good user experience, and it also leads to all sorts of confusion with the implicit differences between the information available at "parse time" versus "runtime." (If you've never had to dig into the docs on pre/post hook rendering, and double-curly nesting, that's probably a good thing: https://github.com/dbt-labs/docs.getdbt.com/issues/2818)

> Allow docs blocks in sql files, with possible support to auto-attach them to models (since they appear in the model sql file anyway)

This specific idea (#979) got quite a bit of traction. As you say, I'm not sure how we'd do the "auto-attaching," except via some implicit magic with the naming convention (docs block named `model_name__column_name`, or named `column_name` and defined in a model file named `model_name`).

> Custom "javadoc"-style annotations, possibly surrounded by /** */ blocks (which can help with IDE confusion around jinja blocks)

This would require us to do SQL parsing, which isn't something dbt has done to date. While the answer is definitely not "never," I'm quite wary of taking an accidental half-step into this particular quagmire.

It would be much more tenable for Python models, where we've already toyed with the idea of supporting model-level descriptions via docstring:
```python
def model(dbt, session):
    """
    This is my awesome model!
    """
    df = ...
```

## What about YFM?

Here's an idea, which you have every right to hate: What if we did this as **yaml front matter (YFM)**? After scanning through the older linked issues, I didn't see this come up; I may have missed it, or I may be proposing something that everyone else has tacitly agreed would be terrible. For whatever reason, I don't hate the idea, at least not at first glance.

Idea being, these remain two separate & independent sets of information, which just happen to live in the same file. The yaml in the front matter can include any of the key-value pairs that could also be passed to the `models:` list in a yaml file. It would be parsed at the same time as the rest of the model file, but using our yaml parser, instead of the `ModelParser`.

The front matter would either take priority over, or be mutually exclusive with, defining yaml properties for this model in a separate yaml file. If the model also defines `{{ config() }}` (Jinja-SQL) or `dbt.config()` (Python), it would take precedence over, or be mutually exclusive with, the config defined in the front matter.
```sql
---
description: This is my model's description. It can reference a {{ doc('block') }} if it wants to.
config:
  materialized: incremental
  unique_key: id
columns:
  - name: first_col
    tests:
      - unique
      - not_null
---

select ...
```
```py
---
columns:
  - name: first_col
    tests:
      - unique
      - not_null
---

def model(dbt, session):
    ...
```

> The model file might get too big (dozens of columns in a model might be confusing, especially if all the docs are inlined)

This approach definitely doesn't help with the sheer model size, but yaml does offer a few line-saving options if you're willing to get a bit JSON-y:
```sql
---
columns:
  - {name: first_col, tests: [not_null, unique]}
  - {name: second_col, tests: [not_null], description: ...}
  - ...
---

select ...
```
To be clear, what I'm **not** proposing here is the ability to use Jinja or yaml to dynamically create/populate the front matter contents. I do think that remains, rightly, the domain of code-generation plugins/packages. (Our work on refactoring `dbt-core`'s foundational APIs could eventually unlock even more-capable code generation, in the form of more-direct _node_ generation.)

**Reply by jakebiesinger** [reply]
@jtcohen6 Thanks so much for the quick response here!

> I couldn't promise a walk in the park.

That's fine. This feature seems worth the effort to me. Thanks in advance for your patience with what I'm sure will be missteps on my part in the implementation :) 

As for the YAML Front Matter idea, I love it. 😍 

DBT is heavy on the YAML anyway and this would save us from having to define any weird semantics (like javadoc) and from doing ugly expansions to `config`, which seem to have their own shortcomings. Additionally, this becomes a straightforward extension of the `schema.yml` files. There are no new ideas to explain aside from "another place you can specify model schema items is in the model itself, as YAML Front Matter".  I like the idea that this is equivalent to specifying these YAML blocks inside the model in a schema.yml -- it handles the "auto-attach to the model" business nicely.

FWIW, this is probably enough for me to go on to at least produce a proof-of-concept. I'll see if I can sneak in some time this week for it, unless you'd prefer to hear more ideas from the community... but this seems very clean to me.

**Reply by jakebiesinger-storyhealth** [reply]
@jtcohen6 I've played around with the code a bit and I'm close to having a working POC. Some observations ... let me know what you think!

I'm concerned that YAML Front Matter will be too easy to accidentally specify in a SQL-heavy world. For example, `--` is a comment character in many SQL dialects so existing model files are just one or two extraneous `-`'s away from having a YFM.

A few options here include:
  - Guard this new behavior behind a flag, say `enable_schema_definitions_in_model_files: true` though I'm not sure where to put it. `profiles.yml` seems to be the place for other flags, but this option feels more closely aligned to the project itself than the user who's running `dbt` or the environment it's being deployed to. This could be an optional, top-level entry in `dbt_project.yml`. Are there implications I should know about w.r.t. separate packages (and those packages ability to use this incoming feature)?
  - Use a custom "frontmatter" header + footer. I'm looking at [python-frontmatter](https://python-frontmatter.readthedocs.io/en/latest/) and it allows custom delimiters. The YFM `---` doesn't play nicely with IDEs (will need formatter + plugin updates to make this `jinja-sql-yaml`) and we'd need also need to strip the frontmatter from the rendered SQL. 

So! What if we hack this a bit with custom delimiters, and make it specific with a special tag like `dbt-model`? 
 
Compare...

YAML frontmatter w/`---`
<img width="352" alt="image" src="https://user-images.githubusercontent.com/66106280/218332949-94b0254e-313c-4977-ac74-b6ff55f19883.png">

vs
javadoc w/ `dbt-model` style
<img width="359" alt="image" src="https://user-images.githubusercontent.com/66106280/218332973-8a7aac29-1ae7-4640-bca5-3813000b7580.png">

vs 
Jinja comments w/`dbt-model`
<img width="314" alt="image" src="https://user-images.githubusercontent.com/66106280/218333289-4f6ef1d0-22a2-4736-97aa-f665196586f5.png">

It feels like classic `---` YFM is more likely to be understood given its usage in other projects, but feels the most susceptible to errors and like the most downstream work for parsers + plugins. 

I'm currently leaning toward the Jinja-comment style. The block would be stripped from the rendered SQL automatically (a nice shortcut) and still wouldn't induce us to actually use Jinja to parse the schema. @jtcohen6 would you still have the same hesitation for Jinja-that's-not-parsed-as-jinja?

**Reply by jakebiesinger-storyhealth** [reply]
@jtcohen6 WIP PR is up (opening against my own fork, seeking early feedback before I open a proper PR against the `main` branch here) https://github.com/jakebiesinger/dbt-core/pull/1

**Reply by jtcohen6** [reply]
@jakebiesinger Props for making this happen so quickly!! And really appreciate your eye toward the potential for gnarly UX, given the similarity between YFM and SQL comments (`--`).

I've been jamming on this idea in a slack thread with @dbeatty10 @gwenwindflower @joellabes @callum-mcdata (guardians of the guardrails, defenders of the dbt-onic, connoisseurs du meilleur conseil). I'll quickly summarize some of what we've agreed on:
- We like choosing for YFM as a standard, rather than opting for something more custom (javadoc-inspired). It would be possible to build other workflow / code-gen tools that integrate with YFM, which could be cool!
- While the distinction between `--` and `---` is quite subtle, I think we can have simple rules to avoid YFM <> SQL comment collision:
    - YFM must be located at the top of a file. That is, the first character in the (whitespace-stripped) file contents must be `---`. (I think it's okay to prohibit comments above YFM.)
    - We `try` to parse the contents (between start `---` and end `---`) as valid yaml, and just skip/ignore it if it isn't, rather than raising an exception. We might still raise a warning: "You have something that _looks like_ YFM, but it isn't! Is it meant to be?"
- We share an initial hesitation about adding _yet another way to configure models_. There should be One Right Way to do things. **AND**, we actually _like_ YFM a lot more than `{{ config() }}`, which has a wonky syntax and confusing behavior around when its contents are rendered. This could be our way of slowly-but-surely moving away from `{{ config() }}` as the recommended/standard pattern for in-file configuration. There will be some things that `{{ config() }}` can do, which YFM can't: accepting macros as inputs, and late-rendering hooks. Those are known limitations of yaml configs, and we can solve for them over time; it's out of scope for this work.
- Should it be possible to define a custom `name` for a model within the in-file YFM, and thereby enable a model's name to differ from the name of its file? It's a neat idea, but worth thinking through all the implications before implementing. Let's call it out of scope for now.

**Reply by jakebiesinger** [reply]
@jtcohen6 Well, it's a pleasure to have this issue included with such auspicious company... 👋  howdy folks!

Cool, this sounds like a good direction to take, and cool that you're thinking of future possibilities and even of moving away from `config`. I also think the strictness of this being the first non-whitespace entry in the file will go a long way toward preventing overlap with SQL comments. People are still going to run into that problem occasionally, but as you say, we can add some warnings during parsing that "we got some weird non-YAML stuff inside the YFM". This also means the various downstream jinja-sql parsers will need updates, but if this is the long-term direction you want, it seems totally acceptable.

Mechanically, this means we'll need to also update the model parsing code to remove the successfully-parsed YFM from the "raw" file content-- the other solutions don't have to worry about that problem.

I'll take another stab at updating the PR in the coming days, and will attempt to take care of some of the outstanding TODO's (such as working with partial-parsing properly).

Thanks for the discussion and direction!

**Reply by joellabes** [reply]
> We might still raise a warning: "You have something that looks like YFM, but it isn't! Is it meant to be?"

It is correct for this to be a warning. That said, by the power vested in me by #6165, I would elevate it to be an error in _my_ projects.

> the first character in the (whitespace-stripped) file contents must be `---`. (I think it's okay to prohibit comments above YFM.)

Would it be `---\n`, or would we also try to parse 
```
-------------------------------------------------------------------------
----------------------- NOTE TO FUTURE DEVELOPERS: ----------------------
------------------- REMEMBER TO CHANGE THE OTHER FILE -------------------
-------------------------------------------------------------------------

with src as (
select * from {{ source('blah', 'blah') }}
),
...
```
?

**Reply by jakebiesinger** [reply]
@joellabes how's about `---\s*`, meaning `---` followed by any amount of whitespace?

**Reply by joellabes** [reply]
if it's validated in a regex, I'd probably advocate for `^---$` 

Although as I type the dreaded R word, it strikes me that that there's presumably some sort of YFM-parser already available in Python that has already handled all these edge cases?

**Reply by jakebiesinger** [reply]
There is https://pypi.org/project/python-frontmatter/ which I'm using in my PR above, but I looked around at the internals and it's literally just [regex parsing](https://github.com/eyeseast/python-frontmatter/blob/main/frontmatter/default_handlers.py#L233) combined with a `documentString.split(regex)`. This unfortunately accepts frontmatter from anywhere in the doc (not something we can configure, even by implementing a custom handler in that world. While there aren't many deps for that package, the latest release is March 2021...

Honestly, I think I'd rather provide our own stricter and simpler parser, given the above's flexibility + complexity. A few dozen lines of code is not much of a price to pay for parsing exactly how you want IMHO.

**Reply by jakebiesinger** [reply]
@jtcohen6 https://github.com/jakebiesinger/dbt-core/pull/1 now has an implementation of the YAML Frontmatter parser as described. 

I am working off of the assumption that
1. Leading whitespace (in any quantity) is fine
2. Whitespace trailing the `---` in the frontmatter header or footer is fine
3. The header + footer lines must start with the `---` (no preceding whitespace or other text)
4. YAML doc blocks beyond the first will not be parsed (and will be included in the compiled model code)

@joellabes let me know if point 2. is too lenient for your tastes (and we should use `^---$` instead).

Some weirdness to report in this implementation:
1. The requirement that failing to parse the YAML block should cause it to remain in the compiled model feels a little weird, and in this implementation necessitates us to parse the file (including the embedded schema stuff) twice.
2. Since .sql files can now be the source of tests, the compiled output can't use those files as parent directories, necessitating a `subnodes` sibling folder.
3. I didn't put much effort into partial parsing. I started to go down that path and realized that our model files now want to be both `SourceFile`s and `SchemaSourceFile`s, but the semantics for partial parsing those are pretty different. I opted to just keep it simple for now instead of trying to model some multiple inheritance of the two.

This implementation also inherits the behavior where a given model can only be configured by one `schema.yml` file. In this implementation, that means that `schema.yml` files can't specify anything about models that also have YAML frontmatter in them.

The integration + unit test battery is passing; I need to write up some more tests for new behavior and do changie things, but I think this is approaching review readiness.

**Reply by joellabes** [reply]
I like all 4 of your decisions for the implementation of YFM identification! 

> The requirement that failing to parse the YAML block should cause it to remain in the compiled model feels a little weird, and in this implementation necessitates us to parse the file (including the embedded schema stuff) twice.

Since we're being pretty strict about what we consider YFM (needs to be two sets of totally empty --- lines), it probably is reasonable to reject it outright as bad YAML instead of consider it a funny sql comment. 

>In this implementation, that means that schema.yml files can't specify anything about models that also have YAML frontmatter in them.

I think this is somewhere between an acceptable tradeoff and actually great news! It feels good to me that there is not a new layer of inheritance and override order of operations for people to get their heads around. 

:point_up: This is provided that if they try, they get a warning that some properties are being ignored, and the winner is chosen in a deterministic way. I don't really care which one wins if one is technically easier than the other, but all else being equal I'd vote for YFM being prioritised over schema.yml

I have no comment on the specific implementation details you've run into! I will leave that for the people who know what they're doing 😂

**Reply by vergenzt** [reply]
Love that this issue was getting traction! One question I had about the YFM approach -- will Jinja templating be applied to the whole file before trying to parse don't matter out, or does templating get applied only to the SQL body?

I'm hoping the former -- I'd love to use macros to help me scaffold out the YFM!

One of the nice aspects of the "Javadoc"-esque approach that I liked was that I could use macro to define all parts of a column --implementation, column name, *and* description.

**Reply by joellabes** [reply]
[Jinja would only be available in the SQL body](https://github.com/dbt-labs/dbt-core/discussions/6853#discussioncomment-4995554)
> There will be some things that `{{ config() }}` can do, which YFM can't: accepting macros as inputs, and late-rendering hooks. Those are known limitations of yaml configs, and we can solve for them over time; it's out of scope for this work.

Getting Jinja in the YFM probably looks something like [this](https://github.com/dbt-labs/dbt-core/discussions/6170#discussioncomment-3993081), in pursuit of a day where `{{ config() }}` and its assorted baggage isn't necessary

**Reply by jakebiesinger** [reply]
Just checking in here... I've made a lot of progress but have had to scrap my approach 2-3 times. Getting this to work for a full parse is pretty trivial, but partial parsing is a beast. The PR has been updated with latest progress. I definitely don't understand enough about how this is supposed to work, so a lot of what I'm doing feels like shooting in the dark.

Current status:
- Seems to work 100% for full reparse
- In partial-parsing mode, the saved manifest doesn't seem to be retaining the changes from YFM

There are a lot of things I've tried so far, but I wonder if it would be helpful at this point to connect to someone who understands this code, just so I could get some pointers... say 15 minutes? @joellabes or @jtcohen6 is something like that possible? Or if there's a good Slack channel I can ask my questions and explain my approach, I'd be all for it. I'm certain I'm doing some silly things that another set of eyes would clear up quickly.

**Reply by joellabes** [reply]
Sorry to hear it's been a mission! I'm not a Python dev so won't be much help with your implementation woes, but the best first step is probably to open a draft PR containing your code and do a self-review, indicating all the places you'd like feedback and the sorts of issues you're running into, so that someone from the Core team can give more directed help

(Since you asked, there's also a #dbt-core-development channel in the Slack)

**Reply by jtcohen6** [reply]
@jakebiesinger I'd also be open to grabbing some time (~30 min) to chat! I'll send you a DM in the dbt Community Slack; the next few weeks are busy, but we'll see if we can find / make something work.

@gshank has offered to share some wisdom on partial parsing. As Joel recommended, if you could push up a draft PR and call out the problem spots in the code, she'll be able to take a look before we chat. It's possible that the codepaths here are tangled enough, that our best bet will be to agree on a reasonable stopping point, and for someone on our side to carry it over the finish line. (Even if that's the way we choose to go, we'd never have gotten this far if you hadn't first taken the initiative!)

FWIW - this discussion has triggered some valuable internal conversation on our team around configs vs. properties — what's defined where, should these really (finally) just be the same thing. The question for us to answer is: If anything could be passed into `{{ config() }}` — including documentation and tests, the original prompt for this issue — is there still value in pursuing the yaml frontmatter approach? I think yes! That's the discussion we had above, before we moved into implementation, and I still stand by it. I think YFM could offer a better user experience than the code-as-config story we currently tell with Jinja macros today, even if the latter will still offer (might always offer) greater flexibility for more-advanced use cases.

**Reply by jakebiesinger** [reply]
@jtcohen6 @joellabes thanks for the pointers and discussion! Next time, I'll start with asking around in Slack. The codebase has definitely gotten more complex since I last touched it 4-5 years ago. I also somehow missed the blurb about you core-development slack channel over in your [CONTRIBUTING](https://github.com/dbt-labs/dbt-core/blob/HEAD/CONTRIBUTING.md#about-this-document) page.

**Comment by gshank** [comment]
@jakebiesinger I took a look at your pull request, and I admire your enthusiasm for diving into some hairy code. However, the approach of creating two different file objects was not the way to go. Once I started thinking about it, it was hard to let it sit, so I borrowed some of your code from yaml_helpers.py and created a draft pull request which calls the model schema parser from the model parser: https://github.com/dbt-labs/dbt-core/pull/7100.

There's still a fair amount of work to do to make sure that edge cases are handled and dealing with config in both the SQL files and a schema file, etc, but I think as a proof of concept it looks pretty reasonable.

Thanks for providing the inspiration :) -- we'll be sure to put your name on as co-contributor.

**Reply by jakebiesinger** [reply]
@gshank thanks for taking this over. I'm just coming back into this community + code after a long hiatus. I probably should have started asking over in the #dbt-core-development Slack channel or talked through approaches, etc. There's only so much you can do with limited time, and I'm just grateful the work is getting some traction.

**Comment by raphael-boegel-pcg** [comment]
For proper code highlighting your IDE propably requires a Plugin.
And if a plugin is required anyway, maybe the hole thing could be solved by your IDE showing both the model and its documentation next to each other. Wouldn't that be a great feature for the dbt cloud IDE to start with?

**Comment by zqureshi** [comment]
When we start evaluating this option we should consider compatibility with existing auto-completion and syntax highlighting in popular code editors and to not accidentally break them.

**Comment by pSpitzner** [comment]
Hi all,

for everyone interested in a temporary workaround, I have hacked together something in [my comment](https://github.com/dbt-labs/dbt-core/issues/11837#issuecomment-4003442996) in #11837

TLDR:
Its a python snippet that scans all sql files for a jinja doc section (via a given convention) and creates markdown + yml files, which get along nicely with `dbt docs generate`

Note: For now, it only takes care of model descriptions.

**Reply by jwehn02** [reply]
@pSpitzner 

> Hi all,
> 
> for everyone interested in a temporary workaround, I have hacked together something in [my comment](https://github.com/dbt-labs/dbt-core/issues/11837#issuecomment-4003442996) in #11837
> 
> TLDR: Its a python snippet that scans all sql files for a jinja doc section (via a given convention) and creates markdown + yml files, which get along nicely with `dbt docs generate`
> 
> Note: For now, it only takes care of model descriptions.

I'm working on a similar workaround using a python script that will parse our .sql model file we use to materialize our semantic schema views in Snowflake.  The script parses the .sql model files, separates them by domain, and then produces static HTML pages that we host on GitHub pages.  But I like the idea of having it produce yml files instead. Does your python code run on a schedule to keep the yml files up to date or is it just triggered on a PR commit?

**Reply by pSpitzner** [reply]
So far we dont use in production. I run it manually, but I guess turning it into a python package and then having it run via github action should not be more than an afternoon's work.

Concerning the yml files: it can update the `_docstrings.yml` it generated in a previous run. Note, however, that the limitation of dbt remains, that multiple yml files per model are not allowed. And my script does not handle the case yet, when another yaml already exists for a model. (One could/should improve it to deal with that case by placing the reference to the markdown files in the exising model yamls.)

**Reply by joshuataylor** [reply]
When dealing with YAML, in python, use ruamel - it's absolutely fantastic, handles comments and can patch weird YAML files perfectly.

It's a ... tad obtuse ... to deal with, if you'd like I can share what I do - but check the dbt autofix PRs for my PR which adds this as an example.

Feel free to ping me if you get stuck.

---

I've looked deep into the void of parsing dbt projects with YAML and manifest.json and will gladly help anyone avoid the same fate 😂
